Pull

phatblat opened the pull request:
Initial Pull implementation preferring fast-forward merge if possible or a "normal" merge if branches have diverged, depending on the merge analysis. Pull will fail if a conflict is detected.

### TODO

* [x] Fast-forward merge
* [x] Normal merge
* [x] Merge conflict communication to caller (via error)
* [x] Tests

### Out of scope
 
* Pull with rebase
* Explicit merge commit (aka `--no-ff`)
* Prevent merge commit (aka `--ff-only`)
* Pull an unborn branch

pietbrauer commented:
Slightly more elegant than what I had: https://gist.github.com/pietbrauer/b1b3d2c6bd927ac85094

Things with open question marks for me:

* [ ] How to fast forward (shell out to libgit2?) to avoid merge commit
* [ ] How to support rebase (I guess libgit2 needs rebase support first?)
* [ ] How to deal with merge conflicts (or how to communicate them back to the caller)

pietbrauer commented:
I also end up with a dirty working copy when trying to pull in 2 commits from the remote:

Phone is at https://github.com/pietbrauer/git2go-test-repo/commit/02ed2a336547592cc2834fc4aef2b744f67fa24e
Remote is at https://github.com/pietbrauer/git2go-test-repo/commit/7c17c5815471e1eef41eae435999f9ea2d4dcc67
```bash
diff --git a/Test.txt b/Test.txt
index 5948f52..34f55f5 100644
--- a/Test.txt
+++ b/Test.txt
@@ -26,5 +26,3 @@ Test 25
 Test 26
 Test 27
 Test 28
-Test 29
-Test 30
```

Can you confirm?

palmin commented:
My insights are for raw libgit2 but I will outline how I merge and perhaps this can be applied in objective-git.

For me ```git_merge_analysis``` is key to getting good merges in the second stop of a push.

I start out getting merge heads for HEAD and corresponding head on remote tracking branch. These merge heads are ```git_annotated_commit``` in libgit2. Then I use ```git_merge_analysis``` to determine if fast-forward is possible, if HEAD is unborn (without any previous commits) or if we need to perform a regular merge. 

fast-forward and unborn HEAD is fixed by pointing to the head commit of the remote tracking branch with ```git_reference_set_target``` but the real work lies in regular merge of course.

The regular merge changes the tree and I use ```git_index_has_conflicts``` to check for conflicts. When there are none I can simply make the merge commit having the old HEAD and remote tracking head commits as parents. 

If there are conflicts these need to be resolved by the user somehow and ```git_index_remove_bypath``` followed by ```git_index_add_bypath``` is used to mark a file as resolved and when everything is resolved a commit can proceed.

Hope this helps in some small way. 

Working Copy does not currently do rebase but it should be a matter of applying ```git_cherrypick``` on the commits that the remote tracking branch is ahead.

pietbrauer commented:
@palmin Thanks so much for sharing your insights!

@phatblat I added `GitMergeAnalysis` support and will try to add support for Fast-Forward.

tiennou commented on ObjectiveGit/GTRepository+Pull.m:
```diff
@@ -0,0 +1,118 @@
+
+	// Check if merge is necessary
+	GTBranch *localBranch = [repo currentBranchWithError:error];
+	if (*error) {
```
I think you should be checking for the return value, not the error — I fear a nice SEGV if I was to pass `NULL` here. There are other cases below.

phatblat replied:
Good catch. Using the user-supplied error pointer for control flow is messy, and dangerous :boom: .

tiennou commented on ObjectiveGit/GTRepository+Pull.m:
```diff
@@ -0,0 +1,118 @@
+
+- (GTMergeAnalysis)analyseMerge:(GTBranch *)toBranch fromBranch:(GTBranch *)fromBranch error:(NSError **)error
+{
+	git_merge_analysis_t analysis;
```
`toBranch` looks unused ? (I do realize it's WIP, feel free to ignore that if you're still 'writing').

Do note that I find the `GTMergeAnalysis` as a return value strange. I mean, how can we tell between `GTMergeAnalysisNone`, or "some kind of convoluted error occurred". Usually this means that you'll need to pass a pointer to a caller-supplied `GTMergeAnalysis` like so :
`-(BOOL)analyseMerge:(GTMergeAnalysis *)result fromBranch:f toBranch:t error:...`.

pietbrauer replied:
Yep, agreed. `toBranch` is unused and I will remove it. `git_merge_analysis()` works with current branch AFAIK. I will also convert it to your suggested signature as it looks more like the existing APIs although `GITMergeAnalysisNone` is not in use currently see: https://github.com/libgit2/libgit2/blob/master/include/git2/merge.h#L273-L274

tiennou replied:
FYI (because you had me wondering) : https://github.com/libgit2/libgit2/blob/master/src/merge.c#L2672. Not that it really matters though, it's just a default value.

pietbrauer replied:
I see, I was just reading the docks, not looking into the code. Then I will probably make a PR for `libgit2` to change the docs. Nevertheless it is safer to handle `GITMergeAnalysisNone` than omitting it. :wink:

phatblat commented:
Wow, great info @palmin! Thanks!

pietbrauer commented:
@phatblat I just added a draft for unborn and fast forward merges. Seems to work. Although I end up with staged changes, the history seems to be the same as on the remote. success_kid.gif

palmin commented:
Regarding changes after using fast-forward ```git_reference_set_target```. I need to use ```git_checkout_head``` to get the working copy in order.

pietbrauer commented:
@palmin Thanks once again. Done so in https://github.com/libgit2/objective-git/commit/1a2fd36c6892aea0d2379dbb89009679de2fb0e0

phatblat commented:
@pietbrauer did you create a new `referenceByUpdatingTarget:committer:message:error:` method on `GTReference`? The logic after `analyseMerge` gets called is looking for that method and not finding it.

phatblat commented:
If libgit2 doesn't support rebase then I think it's clear that's out of scope for this PR.

pietbrauer commented:
@phatblat Ups, looks like I used `objective-git ~> 0.6.2` in my project. I updated the implementation in https://github.com/libgit2/objective-git/commit/0bd6b02ffa83366d67dc6da0ddd99e37b532cfef

What else is missing now?

phatblat commented:
This `pullBranch:` API isn't going to work in the case of an unborn HEAD since you can't get a `GTBranch` for it. Perhaps with a `GTReference`. In the spirit of moving this forward I'm thinking we leave this case unsupported for now.

phatblat commented:
This is ready for review now. The basic functionality is done and primary cases have tests.

The way this works is:
1. Update the tracking branch using fetch
2. Do a merge analysis to figure out if a fast-forward is possible
3. Merge the local and tracking branch by either:
  * Updating the local branch to the HEAD of the tracking branch (fast-forward)
  * Creating a new merge commit using the merged tree from the two branches

Some things to consider:
* API naming: `-pullBranch:fromRemote:withOptions:error:progress:`
* Unborn branch is not supported because it's not currently possible to get a `GTBranch` for one. This may be possible using a `GTReference` but as this API requires a `GTBranch`, it should be somewhat apparent that the unborn case is not yet supported.
* The `progressBlock` parameter is only used for the fetch progress. Currently, no progress block is passed to libgit2 for the checkout at the end.
* `ancestorTree` is nil when passed to the `merge:ancestor:error:` method. This works for the test, but may be necessary in more complex scenarios. Perhaps `-analyseMerge:fromBranch:error:` should return the `git_annotated_commit`?
* Reflog messages. Not sure if there's much of a standard format. I see various git clients using different formats.
* I figure pull w/ rebase along with explicit merge commit and fast-forward only will be handled in the future using the `options` dictionary

Unrelated changes in this PR:
* Added/improved `-description` methods to `GTRemote` & `GTRepository`
* Added libgit2 folder reference to workspace (helps greatly with debugging into the library)

joshaber commented on ObjectiveGit/GTRepository+Pull.h:
```diff
@@ -0,0 +1,56 @@
+///
+/// Returns YES if the analysis was successful, NO otherwise (and `error`, if provided,
+/// will point to an error describing what happened).
+- (BOOL)analyseMerge:(GTMergeAnalysis *)analysis fromBranch:(GTBranch *)fromBranch error:(NSError **)error;
```
s/analyse/analyze because :us: spellings.

joshaber commented on ObjectiveGit/GTRepository+Pull.h:
```diff
@@ -0,0 +1,56 @@
+
+/// Analyse which merge to perform
+///
+/// analysis        - The resulting analysis.
```
Nit: these don't need to be indented quite so far.

joshaber commented on ObjectiveGit/GTRepository+Pull.h:
```diff
@@ -0,0 +1,56 @@
+/// Analyse which merge to perform
+///
+/// analysis        - The resulting analysis.
+/// fromBranch      - The remote to pull from.
```
s/pull/merge?

joshaber commented on ObjectiveGit/GTRepository+Pull.m:
```diff
@@ -0,0 +1,170 @@
+#pragma mark - Pull
+
+- (BOOL)pullBranch:(GTBranch *)branch fromRemote:(GTRemote *)remote withOptions:(NSDictionary *)options error:(NSError **)error progress:(GTRemoteFetchTransferProgressBlock)progressBlock
+{
```
Nit: opening curly on the preceding line.

joshaber commented on ObjectiveGit/GTRepository+Pull.m:
```diff
@@ -0,0 +1,170 @@
+
+- (BOOL)pullBranch:(GTBranch *)branch fromRemote:(GTRemote *)remote withOptions:(NSDictionary *)options error:(NSError **)error progress:(GTRemoteFetchTransferProgressBlock)progressBlock
+{
+	NSParameterAssert(branch);
```
Nit: explicit comparison to `nil`.

joshaber commented:
Awesome :metal: 

Any idea why the Travis build is failing?

phatblat commented:
Thanks for the feedback.

The Travis build is failing when trying to run the [Mac tests](https://travis-ci.org/libgit2/objective-git/builds/71580332#L978)

```
  [Info] Collecting info for testables... (1420 ms)
  run-test ObjectiveGit-MacTests.xctest (macosx10.10, iPhone 5s, logic-test)
━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
Failed to query the list of test cases in the test bundle: Error while parsing JSON: The data is not in the correct format.: {
    stderr = "";
    stdout = "Pending: into an empty repo\n[\"GTBlameSpec...
```

Looks like this is an xctool issue, but the issues I found (facebook/xctool#172) are quite old.

Actually, the `run-test` line above looks like it's trying to run the tests on `macosx10.10` but in the `iPhone 5s` simulator. Maybe we need to configure the `cibuild` to pass mac as the platform?

pietbrauer commented:
@phatblat The core problem is that travis aborts builds if they take longer than 10 minutes and produce no output. That changed recently (that is why we added Travis again). One possible workaroung would be to trigger `./script/update_libssh2` manually before starting `xctool`. The reason is because the scripts produce some output while running but `xctool` is omitting the output, until the script finished (which is after about 11 minutes).

phatblat commented:
Ah. Those library build scripts could also be changed to inline the output instead of dumping it into architecture-specific log files (perhaps with an optional verbose flag).

Those log files have been troubling me since they make it hard to watch the build as it runs. Also, libgit2 build outputs hundreds of warnings for the 32-bit arm architectures (not sure how bit an issue that is), but that's not visible unless the open the log since it's all tucked away out of sight. I realize there's a lot of output from those builds, but Xcode hides it all as soon as that build step completes successfully.

phatblat commented:
Merged a handful of changes from master mostly just to kick off another build

modocache commented on ObjectiveGitTests/GTRepository+PullSpec.m:
```diff
@@ -0,0 +1,266 @@
+		});
+
+		/// Conflict During Merge
+		it(@"fails to merge whene there is a conflict", ^{
```
`s/whene/when/g`

phatblat commented:
Oops. Thx @modocache!

I guess it wasn't the :fire: that was causing the `Failed to query the list of test cases in the test bundle: Error while parsing JSON: The data is not in the correct format.` error on Travis. It's got to be something in the tests in the `/pull` branch because our other PRs aren't having this problem.

phatblat commented:
The build for this PR is finally green! Turns out disabling a test using `pending` causes Travis to choke.

Ready for review.

joshaber commented on ObjectiveGitTests/GTRepository+PullSpec.m:
```diff
@@ -0,0 +1,268 @@
+			});
+		});
+
+/* pending tests break build on travis
```
lol :sob:

joshaber commented:
Sorry I'm a scumbag and waited too long to review this :disappointed:. It looks good other than needing to be updated.

phatblat commented:
No worries! I've updated from master. I'd like to request a pull of the Pull pull request. :laughing:

joshaber commented:
>I'd like to request a pull of the Pull pull request. 

![](http://www.reactiongifs.com/r/dkg.gif)

joshaber commented:
:metal:

dleehr commented:
Awesome! :tada:
